Switch managed AlertManager to always-on, and use secrets

damemi opened the pull request:
This updates the managed AlertManager to be always on by default, to run in the `gmp-system` namespace, and to mirror user config from a Secret they create in `gmp-public`.

damemi commented:
Still need to test this, but opening now to show the changes

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -103,6 +103,30 @@ spec:
               object represents. Servers may infer this from the endpoint the client
               submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
             type: string
+          managedAlertManager:
```
`managed` is implied. I'd recommend `alertManager`. 

Also remember to update description etc and documentation in other places.

I think it should be at the same level as `rules` but weird that both are under `collection`. A better section would be `alerting`.. Ah well too late to change.

damemi replied:
Since there is already an `alertmanagers` field under `rules` for users to configure their own unmanaged AM, I think it is worth the distinction here that we are specifically talking about the "default"/managed alert manager

Also this is at the same level as `rules`, not under `collection`: https://github.com/GoogleCloudPlatform/prometheus-engine/blob/450382f6b3baa484941b0500c98b013830bd0d9c/pkg/operator/apis/monitoring/v1/types.go#L49-L58

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -352,20 +413,18 @@ func (r *operatorConfigReconciler) makeAlertManagerConfigs(ctx context.Context,
+		alertManagers = append(alertManagers, monitoringv1.AlertmanagerEndpoints{
+			Name:      amNamespacedName.Name,
+			Namespace: amNamespacedName.Namespace,
+			Port:      intstr.FromInt(9093),
```
Fine for now but note the possibility that users may edit the alertmanager port in the yaml and break this. Can make it configurable if that use case does come up.

pintohutch replied:
What we should do is actually see if the default AM [_service_ ](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/8e89c873d3bc21f735941b968461c5e48c153d9f/cmd/operator/deploy/operator/12-alertmanager.yaml#L23-L35) exists, and use the port in the retrieved service object directly.

I can do that in a follow-up PR to shake off some of the cobwebs 🙂

pintohutch replied:
https://github.com/GoogleCloudPlatform/prometheus-engine/pull/314

xichen2020 reviewed (commented):
Only skimmed over it and looked fine, will take a second pass.

damemi commented:
Updated with the latest feedback in the design discussions:
* Operator expects a specific config secret name: `gmp-public/managed-alertmanager-config`
* This default is set and shown in the OperatorConfig CRD
* Users can edit the value in the CRD to use a custom secret name

Also added some fixes to the config reloader, and an initContainer to address the same issue as https://github.com/GoogleCloudPlatform/prometheus-engine/pull/287

damemi commented:
@xichen2020 is this good to merge?

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -304,11 +332,6 @@ spec:
                       - port
                       type: object
                     type: array
-                  enableManagedAlertManager:
```
I'm assuming this change is not yet released? Otherwise I think it's a breaking change.

damemi replied:
correct this is not yet released, so it won't be breaking. we got it in time :)

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -103,6 +103,34 @@ spec:
+                    description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+                      TODO: Add other useful fields. apiVersion, kind, uid?'
+                    type: string
+                  optional:
```
nit: IIRC yaml fiels are optional by default, so not sure if this is needed.

damemi replied:
this is generated by kubebuilder

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -103,6 +103,34 @@ spec:
               object represents. Servers may infer this from the endpoint the client
               submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
             type: string
+          managedAlertmanager:
```
A bit odd that this has a single key `config`. Any reason to not just move the properties of `config` up a level and remove the `config` key?

damemi replied:
I like that this leaves room to add more settings for the alertmanager. `config` here is referring to the actual config file, but I could see other settings under `AlertmanagerSpec` as well (replicas, auth..)

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -280,6 +299,68 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
+			Annotations: componentAnnotations(),
+			Labels:      alertmanagerLabels(),
+		},
+		Data: map[string][]byte{"config.yaml": []byte(alertManagerNoOpConfig)},
```
Make `config.yaml` a file-level const and reference it here to simplify refactoring if we need to support arbitrary keys in the future.

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -280,6 +299,68 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		// if the config secret is not found, it may have been manually deleted
```
nit: `If ...`

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -280,6 +299,68 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
+
+	if publicSecret.Data != nil {
+		if len(publicSecret.Data) != 1 {
+			return errors.New("alertmanager config secret must have exactly 1 key")
```
IIUC this will cause the reconciler to reenqueue the request and encounter this error again. Is it worth retrying?

I'm curious about how users can discover this error. Will this be logged in operator logs?

damemi replied:
Yeah, this probably doesn't need to requeue. But it should be brought to the user's attention in the operator logs, so updated to a `logger.Error`

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -280,6 +299,68 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
+		// Alertmanager config could have been created from any file name, so
+		// even though there is only 1 key we have to loop over it to get the data
+		for _, bytes := range publicSecret.Data {
+			secret.Data["config.yaml"] = bytes
```
Reference the const here.

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -342,30 +423,28 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorDeployment(ctx context.Con
-		}
+	amNamespacedName := types.NamespacedName{
+		Namespace: r.opts.OperatorNamespace,
+		Name:      "alertmanager",
```
nit: `NameAlertManager`

xichen2020 reviewed (commented):
Do we have kind tests for this?

xichen2020 commented on pkg/operator/operator_config.go:
```diff
@@ -280,6 +300,69 @@ func (r *operatorConfigReconciler) ensureRuleEvaluatorSecrets(ctx context.Contex
+		Namespace: r.opts.PublicNamespace,
+		Name:      "alertmanager-config",
+	}
+	if spec != nil && len(spec.Config.Name) > 0 {
```
What are your thoughts on returning an error if `len(spec.Config.Key) > 0`? I'm mainly thinking that if a user specifies that, it's better to reject it as an error than silently ignoring it, so if we change the behavior in the future to support arbitrary keys, in the worst case it'll remain not working (if e.g., the key doesn't exist), but that's better than changing from working to not working.

damemi replied:
Key is required in [SecretKeySelector](https://github.com/kubernetes/api/blob/v0.24.4/core/v1/types.go#L2054), so it needs something to be set if we're using that type

xichen2020 replied:
True. Perhaps we can check against `alertmanagerConfigKey` and return an error if it's different, since we are defaulting to that?

damemi replied:
That would require users to always create their configmap with a key that matches `alertmanagerConfigKey`. one of our intentions is to eliminate the number of steps users must get exactly right, so we don't want to do that

xichen2020 replied:
Okay it seems we really don't care about the secret key and anything goes as long as the secret has one key. In that case I think it's probably better to remove the `key` property from the config and rename `name` to something like `singleKeySecret` or `keylessSecret` to make the intent clear. If we need to properly support keys in the future I think we'll need to add another field to avoid breaking backward compatibility.

Wdyt?

xichen2020 reviewed (commented):
Thoughts on adding some kind tests for this?

damemi commented:
> Thoughts on adding some kind tests for this?

I can add some basic e2e in https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/pkg/operator/e2e/main_test.go

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -105,31 +105,15 @@ spec:
-                required:
-                - key
-                type: object
+              configSecret:
```
nit: `singleKeyConfigSecret`.

damemi replied:
do you think it's okay to leave `singleKey` out of the name since that's the only option right now? If we do add a multiKey option later, it would make sense to differentiate in the name then. but for simplicity I think the name is good as-is, and just document that it needs to have only 1 key. wdyt?

xichen2020 replied:
So in the future case we'll provide `configSecret` and `multiKeyConfigSecret`? Sounds fine to me.

damemi replied:
yup that's what I was thinking

xichen2020 commented on manifests/setup.yaml:
```diff
@@ -1076,6 +1076,18 @@ spec:
+              configSecret:
+                description: ConfigSecret refers to the name of a single-key Secret
+                  in the public namespace that holds the managed Alertmanager config
+                  file
```
nit: `file.`

xichen2020 commented on cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:
```diff
@@ -103,6 +103,18 @@ spec:
+              configSecret:
+                description: ConfigSecret refers to the name of a single-key Secret
+                  in the public namespace that holds the managed Alertmanager config
+                  file
```
nit: `file.`

xichen2020 commented on pkg/operator/apis/monitoring/v1/types.go:
```diff
@@ -130,9 +133,14 @@ type ExportFilters struct {
+// Alertmanager instance.
+type ManagedAlertmanagerSpec struct {
+	// ConfigSecret refers to the name of a single-key Secret in the public namespace that
+	// holds the managed Alertmanager config file
```
nit: `file.`

damemi commented:
Squashed and added e2e tests
